duckduckgo/Android
# Improve filename guesses when MIME type provided or dots are in the path

## Description

This PR addresses a few issues related to the filename guessing method used in the Download Manager. To guess the filename, this PR improves usage of the MIME type when a URL path segment matches the MIME type, is ambiguous, or contain dots.

This PR also adds test coverage for when MIME types are provided -- most of the existing logic was written assuming no MIME types, which left a gap in proper filename guesses. Saving the file with a guessed name that is closer to what the user expects is a user experience improvement, and also avoids replacing files with the same name unintentionally.

Below are a few issues and their expected and actual behavior.

### Issue: Not utilizing MIME types appropriately

For a URL such as `https://example.com/files/realFilename.pdf` and a valid MIME type (e.g. `application/pdf`):

- Expected filename: `realFilename.pdf`
- Actual filename: `files.pdf`

### Issue: Not appropriately handling dots in the URL path:

For a URL such as `https://example.com/files.folder/realFilename.pdf` and a valid MIME type (e.g. `aplication/pdf`):

- Expected filename: `realFilename.pdf`
- Actual filename: `files.folder`

For a URL such as `https://example.com/files.folder/realFilename.pdf` and no valid MIME type provided:

- Expected filename: `realFilename.pdf`
- Actual filename: `files.folder`

### Issue: Not appropriately handling ambiguous filenames:

For a URL such as `https://example.com/files/realFilename` and a valid MIME type (e.g. `application/pdf`):

- Expected filename: `realFilename.pdf`
- Actual filename: `files.pdf`

### Steps to test this PR

1. Navigate to a URL which contains a `.` in the path of an attachment

I've created these for testing, which I'll remove if/when this PR is merged:

 - https://d1x9zcshh2aa46.cloudfront.net/abc.def/example.pdf (saves as `abc.def`)
 - https://d1x9zcshh2aa46.cloudfront.net/abc.123/example.pdf (saves as `abc.123`)
 - https://d1x9zcshh2aa46.cloudfront.net/123.abc/example.pdf (saves as `123.abc`)
 - https://d1x9zcshh2aa46.cloudfront.net/123.456/example.pdf (saves as `123.456`)
 - https://d1x9zcshh2aa46.cloudfront.net/abc.pdf/example.pdf (saves as `abc.pdf`)

2. Prior to this PR, none of these files will save as `example.pdf`. With this PR, they'll all save as `example.pdf`. Note that none of these have a `Content-Disposition`, implying they're `inline` and don't have a filename specified.
